File: rust/sdk/src/types/account.rs
use crate::account::AccountId;
use crate::error::M10Error;
use crate::types::TxId;
use crate::TransactionExt;
use m10_protos::sdk;
use m10_protos::sdk::transaction_data::Data;
use m10_protos::sdk::IndexedAccount;
use serde::Serialize;
use serde_with::serde_as;
use std::time::{Duration, SystemTime, UNIX_EPOCH};

#[cfg_attr(feature = "format", derive(parse_display::Display))]
#[cfg_attr(
    feature = "format",
    display("id={id} balance={balance} frozen={frozen} currency={code}({decimals}) limit={balance_limit} issuance={issuance:?}")
)]
#[derive(Clone, Debug, Serialize)]
pub struct Account {
    pub id: AccountId,
    pub balance: u64,
    pub frozen: bool,
    pub code: String,
    pub decimals: u32,
    pub balance_limit: u64,
    pub issuance: Option<Issuance>,
}

#[cfg_attr(feature = "format", derive(parse_display::Display))]
#[cfg_attr(
feature = "format",
display("AccountInfo {{ id={id} parent_id={parent_id} public_name={public_name} currency={code}({decimals}) profile_image_url={profile_image_url} }}")
)]
#[derive(Clone, Debug, Serialize)]
pub struct AccountInfo {
    pub id: AccountId,
    pub parent_id: AccountId,
    pub public_name: String,
    pub profile_image_url: String,
    pub code: String,
    pub decimals: u32,
}

#[cfg_attr(feature = "format", derive(parse_display::Display))]
#[cfg_attr(
    feature = "format",
    display("balance={balance} issuance_accounts={issuance_accounts} holding_accounts={holding_accounts}")
)]
#[derive(Clone, Debug, Serialize)]
pub struct Issuance {
    pub balance: u64,
    pub issuance_accounts: u64,
    pub holding_accounts: u64,
}

#[serde_as]
#[derive(Clone, Debug, Serialize)]
pub struct AccountUpdate {
    pub tx_id: TxId,
    #[serde_as(as = "serde_with::hex::Hex")]
    pub context_id: Vec<u8>,
    pub success: bool,
    pub timestamp: SystemTime,
    pub update_type: AccountUpdateType,
}

#[cfg(feature = "format")]
impl std::fmt::Display for AccountUpdate {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        let Self {
            tx_id,
            context_id,
            success,
            timestamp,
            update_type,
        } = self;
        write!(f, "AccountUpdate{{ tx_id={tx_id} context_id={} success=${success} timestamp={timestamp:?} update_type=${update_type} }}", hex::encode(context_id))
    }
}

#[cfg_attr(feature = "format", derive(parse_display::Display))]
#[derive(Clone, Debug, Serialize)]
pub enum AccountUpdateType {
    #[cfg_attr(
        feature = "format",
        display("SetFrozen{{ account_id={account_id} is_frozen={is_frozen} }}")
    )]
    SetFrozen {
        account_id: AccountId,
        is_frozen: bool,
    },
    #[cfg_attr(
        feature = "format",
        display("SetBalanceLimit{{ account_id={account_id} limit={limit} }}")
    )]
    SetBalanceLimit { account_id: AccountId, limit: u64 },
    #[cfg_attr(
        feature = "format",
        display("SetInstrument{{ account_id={account_id} code={code}({decimals}) }}")
    )]
    SetInstrument {
        account_id: AccountId,
        code: String,
        decimals: u32,
    },
    #[cfg_attr(
        feature = "format",
        display("NewAccount{{ account_id={account_id:?} }}")
    )]
    NewAccount { account_id: Option<AccountId> },
}

impl TryFrom<IndexedAccount> for Account {
    type Error = M10Error;

    fn try_from(account: IndexedAccount) -> Result<Self, Self::Error> {
        let instrument = account.instrument.ok_or(M10Error::InvalidTransaction)?;
        Ok(Self {
            id: AccountId::try_from_be_slice(&account.id)?,
            balance: account.balance,
            frozen: account.frozen,
            code: instrument.code,
            decimals: instrument.decimal_places,
            balance_limit: account.balance_limit,
            issuance: account.issuance.map(Issuance::try_from).transpose()?,
        })
    }
}

impl TryFrom<sdk::indexed_account::Issuance> for Issuance {
    type Error = M10Error;

    fn try_from(issuance: sdk::indexed_account::Issuance) -> Result<Self, Self::Error> {
        Ok(Self {
            balance: issuance.issued_balance,
            issuance_accounts: issuance.non_leaf_children,
            holding_accounts: issuance.leaf_children,
        })
    }
}

impl TryFrom<(&sdk::transaction_data::Data, &[u8])> for AccountUpdateType {
    type Error = M10Error;

    fn try_from(
        (data, account_created): (&sdk::transaction_data::Data, &[u8]),
    ) -> Result<Self, Self::Error> {
        match data {
            Data::CreateLedgerAccount(_) => Ok(AccountUpdateType::NewAccount {
                account_id: AccountId::try_from_be_slice(account_created).ok(),
            }),
            Data::SetFreezeState(freeze) => Ok(AccountUpdateType::SetFrozen {
                is_frozen: freeze.frozen,
                account_id: AccountId::try_from_be_slice(&freeze.account_id)?,
            }),
            Data::SetBalanceLimit(limit) => Ok(AccountUpdateType::SetBalanceLimit {
                limit: limit.balance_limit,
                account_id: AccountId::try_from_be_slice(&limit.account_id)?,
            }),
            Data::SetInstrument(instrument) => Ok(AccountUpdateType::SetInstrument {
                account_id: AccountId::try_from_be_slice(&instrument.account_id)?,
                code: instrument.code.clone(),
                decimals: instrument.decimal_places,
            }),
            _ => Err(M10Error::InvalidTransaction),
        }
    }
}

impl TryFrom<sdk::FinalizedTransaction> for AccountUpdate {
    type Error = M10Error;

    fn try_from(txn: sdk::FinalizedTransaction) -> Result<Self, Self::Error> {
        let response = txn.response.as_ref().ok_or(M10Error::InvalidTransaction)?;
        let context_id = txn
            .request
            .as_ref()
            .ok_or(M10Error::InvalidTransaction)?
            .context_id
            .clone();
        let success = response.error.is_none();
        let timestamp = UNIX_EPOCH + Duration::from_micros(response.timestamp);
        let tx_id = response.tx_id;
        let update_type = AccountUpdateType::try_from((
            txn.data().ok_or(M10Error::InvalidTransaction)?,
            response.account_created.as_slice(),
        ))?;
        Ok(Self {
            tx_id,
            context_id,
            success,
            timestamp,
            update_type,
        })
    }
}

impl TryFrom<sdk::AccountInfo> for AccountInfo {
    type Error = M10Error;

    fn try_from(info: sdk::AccountInfo) -> Result<Self, Self::Error> {
        let id = AccountId::try_from_be_slice(&info.account_id)?;
        Ok(Self {
            id,
            parent_id: if !info.parent_account_id.is_empty() {
                AccountId::try_from_be_slice(&info.parent_account_id)?
            } else {
                id
            },
            public_name: info.public_name,
            profile_image_url: info.profile_image_url,
            code: info.code,
            decimals: info.decimal_places,
        })
    }
}

File: node/m10-sdk/src/types/account.ts
import { m10 } from "../../protobufs";
import { M10Error } from "../error";
import type { TxId } from "../ids";
import { AccountId } from "../ids/account_id";
import { isNone, isSome, unwrap } from "../utils";


export interface IAccount {
    id: AccountId;
    balance: LongNumber;
    frozen: boolean;
    code: string;
    decimals: number;
    balanceLimit: LongNumber;
    issuance: Option<Issuance>;
}

export class Account implements IAccount {
    public id: AccountId;
    public balance: LongNumber;
    public frozen: boolean;
    public code: string;
    public decimals: number;
    public balanceLimit: LongNumber;
    public issuance: Option<Issuance>;


    public constructor(properties: IAccount) {
        this.id = properties.id;
        this.balance = properties.balance;
        this.frozen = properties.frozen;
        this.code = properties.code;
        this.decimals = properties.decimals;
        this.balanceLimit = properties.balanceLimit;
        this.issuance = properties.issuance;
    }

    /**
     * @throws  {M10Error.InvalidTransaction}   if any of the required fields are missing
     */
    public static tryFrom(indexedAccount: m10.sdk.transaction.IIndexedAccount): Account {
        return new Account({
            id: AccountId.fromUint8Array(unwrap(indexedAccount.id, M10Error.InvalidTransaction())),
            balance: unwrap(indexedAccount.balance, M10Error.InvalidTransaction()),
            frozen: unwrap(indexedAccount.frozen, M10Error.InvalidTransaction()),
            code: unwrap(indexedAccount.instrument?.code, M10Error.InvalidTransaction()),
            decimals: unwrap(indexedAccount.instrument?.decimalPlaces, M10Error.InvalidTransaction()),
            balanceLimit: unwrap(indexedAccount.balanceLimit, M10Error.InvalidTransaction()),
            issuance: isSome(indexedAccount.issuance) ? Issuance.tryFrom(indexedAccount.issuance) : null,
        });
    }
}

export interface IIssuance {
    balance: LongNumber;
    issuanceAccounts: LongNumber;
    holdingAccounts: LongNumber;
}

export class Issuance implements IIssuance {
    public balance: LongNumber;
    public issuanceAccounts: LongNumber;
    public holdingAccounts: LongNumber;


    public constructor(properties: IIssuance) {
        this.balance = properties.balance;
        this.issuanceAccounts = properties.issuanceAccounts;
        this.holdingAccounts = properties.holdingAccounts;
    }

    /**
     * @throws  {M10Error.InvalidTransaction}   if any of the required fields are missing
     */
    public static tryFrom(issuance: m10.sdk.transaction.IndexedAccount.IIssuance): Issuance {
        return new Issuance({
            balance: unwrap(issuance.issuedBalance, M10Error.InvalidTransaction()),
            issuanceAccounts: unwrap(issuance.nonLeafChildren, M10Error.InvalidTransaction()),
            holdingAccounts: unwrap(issuance.leafChildren, M10Error.InvalidTransaction()),
        });
    }
}

export interface IAccountInfo {
    id: AccountId;
    parentId: AccountId;
    publicName: string;
    profileImageUrl: string;
    code: string;
    decimals: number;
}

export class AccountInfo implements IAccountInfo {
    public id: AccountId;
    public parentId: AccountId;
    public publicName: string;
    public profileImageUrl: string;
    public code: string;
    public decimals: number;


    public constructor(properties: IAccountInfo) {
        this.id = properties.id;
        this.parentId = properties.parentId;
        this.publicName = properties.publicName;
        this.profileImageUrl = properties.profileImageUrl;
        this.code = properties.code;
        this.decimals = properties.decimals;
    }

    /**
     * @throws  {M10Error.InvalidTransaction}   if any of the required fields are missing
     */
    public static tryFrom(accountInfo: m10.sdk.model.IAccountInfo): AccountInfo {
        const accountId = AccountId.fromUint8Array(unwrap(accountInfo.accountId, M10Error.InvalidTransaction()));
        return new AccountInfo({
            id: accountId,
            parentId: isSome(accountInfo.parentAccountId) ? AccountId.fromUint8Array(accountInfo.parentAccountId) : accountId,
            publicName: unwrap(accountInfo.publicName, M10Error.InvalidTransaction()),
            profileImageUrl: unwrap(accountInfo.profileImageUrl, M10Error.InvalidTransaction()),
            code: unwrap(accountInfo.code, M10Error.InvalidTransaction()),
            decimals: unwrap(accountInfo.decimalPlaces, M10Error.InvalidTransaction()),
        });
    }
}


interface SetFrozen {
    accountId: AccountId;
    isFrozen: boolean;
}

interface SetBalanceLimit {
    accountId: AccountId;
    limit: LongNumber;
}

interface SetInstrument {
    accountId: AccountId;
    code: string;
    decimals: number;
}

interface NewAccount {
    accountId: Option<AccountId>;
}

enum AccountUpdateType {
    SetFrozen,
    SetBalanceLimit,
    SetInstrument,
    NewAccount,
}

type AccountUpdateData = SetFrozen | SetBalanceLimit | SetInstrument | NewAccount;

export interface IAccountUpdate {
    txId: TxId;
    contextId: Uint8Array;
    success: boolean;
    timestamp: LongNumber;
    updateType: AccountUpdateType;
    updateData: AccountUpdateData;
}

export class AccountUpdate implements IAccountUpdate {
    public txId: TxId;
    public contextId: Uint8Array;
    public success: boolean;
    public timestamp: LongNumber;
    public updateType: AccountUpdateType;
    public updateData: AccountUpdateData;

    public constructor(properties: IAccountUpdate) {
        this.txId = properties.txId;
        this.contextId = properties.contextId;
        this.success = properties.success;
        this.timestamp = properties.timestamp;
        this.updateType = properties.updateType;
        this.updateData = properties.updateData;
    }

    /**
     * @throws  {M10Error.InvalidTransaction}   if any of the required fields are missing
     */
    private static getAccountUpdateType(
        txn: m10.sdk.IFinalizedTransaction,
    ): [AccountUpdateType, AccountUpdateData] {
        const txnData = new m10.sdk.transaction.TransactionData(
            unwrap(txn.request?.data, M10Error.InvalidTransaction()),
        );
        switch (txnData.data) {
            case "createLedgerAccount":
                return [
                    AccountUpdateType.NewAccount,
                    {
                        accountId: AccountId.fromUint8Array(
                            unwrap(txn.response?.accountCreated, M10Error.InvalidTransaction()),
                        ),
                    },
                ];
            case "setFreezeState":

                return [
                    AccountUpdateType.SetFrozen,
                    {
                        accountId: AccountId.fromUint8Array(
                            unwrap(txnData.setFreezeState?.accountId, M10Error.InvalidTransaction()),
                        ),
                        isFrozen: unwrap(txnData.setFreezeState?.frozen, M10Error.InvalidTransaction()),
                    },
                ];
            case "setBalanceLimit":
                return [
                    AccountUpdateType.SetBalanceLimit,
                    {
                        accountId: AccountId.fromUint8Array(
                            unwrap(txnData.setBalanceLimit?.accountId, M10Error.InvalidTransaction()),
                        ),
                        limit: unwrap(txnData.setBalanceLimit?.balanceLimit, M10Error.InvalidTransaction()),
                    },
                ];
            case "setInstrument":
                return [
                    AccountUpdateType.SetInstrument,
                    {
                        accountId: AccountId.fromUint8Array(
                            unwrap(txnData.setInstrument?.accountId, M10Error.InvalidTransaction()),
                        ),
                        code: unwrap(txnData.setInstrument?.code, M10Error.InvalidTransaction()),
                        decimals: unwrap(txnData.setInstrument?.decimalPlaces, M10Error.InvalidTransaction()),
                    },
                ];
            default:
                throw M10Error.InvalidTransaction();
        }
    }

    /**
     * @throws  {M10Error.InvalidTransaction}   if any of the required fields are missing
     */
    public static tryFrom(txn: m10.sdk.IFinalizedTransaction): AccountUpdate {
        const [type, data] = AccountUpdate.getAccountUpdateType(txn);
        return new AccountUpdate({
            txId: unwrap(txn.response?.txId, M10Error.InvalidTransaction()),
            contextId: unwrap(txn.request?.contextId, M10Error.InvalidTransaction()),
            success: isNone(txn.response?.error),
            timestamp: unwrap(txn.response?.timestamp, M10Error.InvalidTransaction()),
            updateType: type,
            updateData: data,
        });
    }
}

